lib/uplink/main.go

```diff
@@ -0,0 +1,238 @@
+func (s *Session) GetBucket(ctx context.Context, bucket string) (storj.Bucket,
+	error) {
+	panic("TODO")
+}
+
+type CreateBucketOptions struct {
+	PathCipher Cipher
+	// this differs from storj.CreateBucket's choice of just using storj.Bucket
+	// by not having 2/3 unsettable fields.
+}
+
+// CreateBucket creates a new bucket if authorized
+func (s *Session) CreateBucket(ctx context.Context, bucket string,
+	opts *CreateBucketOptions) (storj.Bucket, error) {
+	panic("TODO")
+}
+
+// DeleteBucket deletes a bucket if authorized
+func (s *Session) DeleteBucket(ctx context.Context, bucket string) error {
+	panic("TODO")
+}
+
+// ListBuckets will list authorized buckets
+func (s *Session) ListBuckets(ctx context.Context, opts storj.BucketListOptions) (
+	storj.BucketList, error) {
+	panic("TODO")
+}
+
+// Share creates a new share, potentially further restricted from the Share used
+// to create this session.
+func (s *Session) Share(ctx context.Context, caveats ...Caveat) (Share, error) {
+	panic("TODO")
+}
+
+// ObjectMeta represents metadata about a specific Object
+type ObjectMeta struct {
+	Bucket   string
+	Path     storj.Path
+	IsPrefix bool
+
+	Metadata map[string]string
+
+	Created  time.Time
+	Modified time.Time
+	Expires  time.Time
+
+	Size     int64
+	Checksum string
+
+	// this differs from storj.Object by not having Version (yet), and not
+	// having a Stream embedded. I'm also not sold on splitting ContentType out
+	// from Metadata but shrugemoji.
+}
+
+// GetObject returns a handle to the data for an object and its metadata, if
+// authorized.
+func (s *Session) GetObject(ctx context.Context, bucket string, path storj.Path) (
+	ranger.Ranger, ObjectMeta, error) {
+	panic("TODO")
+}
+
+// ObjectPutOpts controls options about uploading a new Object, if authorized.
+type ObjectPutOpts struct {
+	Metadata map[string]string
+	Expires  time.Time
+
+	// the satellite should probably tell the uplink what to use for these
+	// per bucket. also these should probably be denormalized and defined here.
+	RS            *storj.RedundancyScheme
+	NodeSelection *miniogw.NodeSelectionConfig
+}
+
+// PutObject uploads a new object, if authorized.
+func (s *Session) PutObject(ctx context.Context, bucket string, path storj.Path,
+	data io.Reader, opts ObjectPutOpts) error {
+	panic("TODO")
+}
+
+// DeleteObject removes an object, if authorized.
+func (s *Session) DeleteObject(ctx context.Context, bucket string,
+	path storj.Path) error {
+	panic("TODO")
+}
+
+type ListObjectsField int
+
+const (
+	ListObjectsMetaNone        ListObjectsField = 0
+	ListObjectsMetaModified    ListObjectsField = 1 << iota
+	ListObjectsMetaExpiration  ListObjectsField = 1 << iota
+	ListObjectsMetaSize        ListObjectsField = 1 << iota
+	ListObjectsMetaChecksum    ListObjectsField = 1 << iota
+	ListObjectsMetaUserDefined ListObjectsField = 1 << iota
+	ListObjectsMetaAll         ListObjectsField = 1 << iota
+)
```

Comment:
I always imagined this as a bitmask, not an array of integers.

Reply:
it's totally a bitmask internally. using the variadic argument style is just a bit more go idiomatic and future proof. we want this to be an unsurprising, easy to use, stable api.

Reply:
oh wait, whoops, my comment makes no sense. i refactored to not even have the variadic argument thing. this is a bitmask in the ListObjectsConfig struct after all

Reply:
If we want a stable API then `ListObjectsMetaAll` should be set to "max int", i.e. all bits set to 1. Otherwise adding a new option will shift `ListObjectsMetaAll`, which would be a backward-incompatible change.

Also, we need to choose either `int32` or `int64`, but not `int`. So the number of bits is deterministic on all platforms.

Reply:
well but the benefit of Go is that as long as people use our types and values, if we change something downstream users will recompile. this code side can change between releases.

ironing out our network protocol for stability i think is another separate issue, if the bitmask is part of that too

Reply:
I thought we have this bitmask to filter the metadata on the server side (satellite), so it's part of the network protocol. If it's only on the client side then I don't see the benefit of having it.

Reply:
okay, i'm just saying the goal of this PR is to worry about how we expose the details to the user. we can still use a bitmask for the network protocol, but that's an unrelated issue. we should probably make tickets to audit, harden, and future proof our network protocol in general